### DISCUSSION TITLE
Local vs. Vercel Production: NextJS 14 server component runs smoothly locally, but crashes in Vercel.

### DISCUSSION THREAD

**Discussion by unclebay143** [vercel/next.js#65203] [App Router] [answered]
I'm stuck with a weird issue! My Next.js 14 app works perfectly on my local machine, but when I deploy it to Vercel, a server component with data fetching throws an error.

![image](https://github.com/vercel/next.js/assets/58919619/135a3b87-1e7c-49d4-b668-5a0cb93ca651)

Here's a breakdown:

- Tech Stack: Next.js 14 app with a server component fetching data from MongoDB.
- Local vs. Production: Runs smoothly locally, but crashes in Vercel.
- Client-Side Works: Client pages with data fetching work fine
- Database: MongoDB - connection is okay.

I've build the project successfully with pnpm build, but the issue persists only on Vercel.
![image](https://github.com/vercel/next.js/assets/58919619/11b640f6-774d-44bb-b660-43ccb8b84085)

I'm looking for some help to identify what might be causing this discrepancy and get my server component working smoothly in production! cc: @balazsorban44 (saw your help [here](https://medium.com/@kaloyan_17221/fix-vercel-next-js-fetch-failed-from-undici-polyfill-8c66346c9c2f))

I can grant access to the private repository on demand.

**Comment by icyJoseph** [comment]
Hi,

Well, do you have logs? Around the places where you query for data from your db, you should handle exceptions, and have detailed logs for them. 

I am also not certain about one thing, you say that the project builds locally, but it also builds on Vercel, right? so the build step is not the issue here, but the runtime. Does it also run locally? with `next start`? After running a `next build` locally?

**Reply by unclebay143** [reply]
Thanks for your response @icyJoseph 

> Well, do you have logs?

Here's one of the error from the log on Vercel

```
TypeError: fetch failed
    at Object.fetch (node:internal/deps/undici/undici:11731:11)
    at async globalThis.fetch (/var/task/.next/server/chunks/8457.js:1:36432)
    at async h (/var/task/.next/server/app/dashboard/hackathons/page.js:66:8652)
    at async h (/var/task/.next/server/app/dashboard/hackathons/page.js:66:6665) {
  cause: RequestContentLengthMismatchError: Request body length does not match content-length header
      at write (node:internal/deps/undici/undici:8590:41)
      at _resume (node:internal/deps/undici/undici:8563:33)
      at resume (node:internal/deps/undici/undici:8459:7)
      at connect (node:internal/deps/undici/undici:8446:7) {
    code: 'UND_ERR_REQ_CONTENT_LENGTH_MISMATCH'
  }
}
```


Here's a typical example of one of the server functions for fetching data (in `server.server.ts`):

```
type GetStudentsResponse = { students: Students };
export async function getStudents(): Promise<GetStudentsResponse | undefined> {
  console.log('Running getCurrentHackathon');

  try {
    const url = `${baseURL}/api/students`;
    const result = await fetch(url, {
      cache: 'no-cache',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!result.ok) return undefined;

    const students = await result.json();
    return students;
  } catch (error) {
    console.log(`Error from getStudents Error:- ${error}`);
  }
}
```

Then I'm using them in the server component like this (also using Promise.all here if they're more than 1, sharing this as my approach as well)

```
const [studentsRes, currentHackathonRes] = await Promise.all([
    getStudents(),
    getCurrentHackathon(),
  ]);
```

> you should handle exceptions, and have detailed logs for them.

Here's the logs from the function as well
<img width="1190" alt="image" src="https://github.com/vercel/next.js/assets/58919619/bf5915f4-daa5-4484-86e6-c8a87f590b14">


> I am also not certain about one thing, you say that the project builds locally, but it also builds on Vercel, right? so the build step is not the issue here, but the runtime. 

Yes it builds on Local and Vercel (deployed), but when accessing any server component/page route or clicking their links, the error shows up.

> Does it also run locally? with next start? After running a next build locally?

Yes, it does

<img width="906" alt="image" src="https://github.com/vercel/next.js/assets/58919619/b2b2f736-92be-4e73-93c2-5e4cffeefe8a">

Let me know if there's anything you'd want me to try or provide. Thanks

**Reply by unclebay143** [reply]
[Quick update]


I realized that one of the server component was actually rendering data, checked the difference and noticed the one working doesn't have `headers`, removed the `headers` from the other one, and it's working...

Doing the same for others to see if it works

**Reply by unclebay143** [reply]
Passing the `headers()` to the fetch method seems to be the issue, but without it some APIs that requires authentication won't work.


```
export async function getAllActivityAudits(): Promise<
  { audits: Audits | [] } | undefined
> {
  try {
    const url = `${baseURL}/api/audits`;
    const result = await fetch(url, {
      // headers: headers(),
      cache: 'force-cache',
    });

    if (!result.ok) return { audits: [] };
    const audits = await result.json();
    return audits;
  } catch (error) {
    console.log(`Error from getAllActivityAudits: Error:- ${error}`);
  }
}
```

The API route requires `session`

```
  const session = await getServerSessionWithAuthOptions();
    if (!session) {
      return NextResponse.json(
        { message: 'Session required' },
        { status: 403 },
      );
    }
```

Am I doing something wrong here. WDYT?

**Reply by icyJoseph** [reply]
Try with `headers: new Headers(headers())` ~ 

Just noticed that the logs show content length disagreement, that might be because headers() gives all headers in the incoming request, which are then used when OP sets up the request to the 3rd party server, and that is bound to cause issues, for example with content length

**Reply by unclebay143** [reply]
Thanks @icyJoseph, I had this custom header previously as part of fixes that I've tried. Didn't seem to resolve it.

```
export const getCustomHeaders = () => {
  const defaultHeaders = headers();
  const customHeaders = new Headers(defaultHeaders);
  return customHeaders;
};
```

What I'm trying now is to move the db calls to the server fetch function itself. Since the server component will run first on the server.

Something like this:

```
type GetAllActivityAuditsResponse = { audits: Audits };
export async function getAllActivityAudits(): Promise<
  GetAllActivityAuditsResponse | undefined
> {
  try {
    const session = await getServerSessionWithAuthOptions();
    const student = await StudentModel.findOne({ email: session?.user.email });
    const audits = await AuditTrailModel.find({ student: student.id }).sort({
      createdAt: -1,
    });

    return { audits: JSON.parse(JSON.stringify(audits)) };
  } catch (error) {
    console.log(`Error from getAllActivityAudits: Error:- ${error}`);
  }
}
```

**Reply by unclebay143** [reply]
Okay, moving the db calls to the function is timing out:
![image](https://github.com/vercel/next.js/assets/58919619/d59b85d7-3296-41b6-9325-106a7ff87c24)

![image](https://github.com/vercel/next.js/assets/58919619/c84bc390-657f-4155-8a5c-6b2a51bb735d)

**Reply by icyJoseph** [reply]
Right, but I guess the issue is that you are passing all incoming headers into the next request, and there's content length headers there as well, among other stuff that might set some expectations for the next request you make.

Pick the headers you want to forward, and build a new headers object with those. Or clone the headers and remove the ones that you don't need, which doesn't seem to be maintainable.

**Reply by unclebay143** [reply]
That seems to make sense to me. However, I'm not sure of which header is needed here, I'm making use of `next-auth` for my authentication and to get the session. Do you by chance have an idea on this?

**Reply by unclebay143** [reply, marked as answer]
Hi @icyJoseph 
I did something like this with the `cookies` and it's working. ref: https://www.reddit.com/r/nextjs/comments/1ajr38s/comment/kp32gvk/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

```
type GetAllActivityAuditsResponse = { audits: Audits };
export async function getAllActivityAudits(): Promise<
  GetAllActivityAuditsResponse | undefined
> {
  const sessionCookie = cookies();
  try {
    const url = `${baseURL}/api/audits`;
    const result = await fetch(url, {
      headers: {
        Cookie: `${sessionCookie}`,
      },
      cache: 'force-cache',
    });

    if (!result.ok) return { audits: [] };
    const audits = await result.json();

    return audits;

  } catch (error) {

    console.log(`Error from getAllActivityAudits: Error:- ${error}`);
  }
}
```

Not sure why it worked 😄, any idea?

[Update]
Works now because server requires cookies for processing request.

**Reply by unclebay143** [reply]
[Update]: Everything seems to be working just fine now with the cookies update
